src/typing/type_inference_js.ml

```diff
@@ -1292,8 +1292,29 @@ and toplevels cx stmts =
       if uc < List.length stmts
       then (
         let msg = "unreachable code" in
-        let (loc, _) = List.nth stmts uc in
-        Flow_js.add_warning cx [mk_reason "" loc, msg]
+        let rec drop n lst =
+          match (n, lst) with
+          | (_, []) -> []
+          | (0, l) -> l
+          | (x, h :: t) -> drop (x-1) t
+        in
+        let trailing = drop uc stmts in
+        trailing |> List.iter Ast.Statement.(fun stmt ->
+          match stmt with
+          (* function declarations are hoisted, so not unreachable *)
+          | (_, FunctionDeclaration {
+                FunctionDeclaration.id;
+                params; defaults; rest;
+                body;
+                returnType;
+                typeParameters;
+                async;
+                _
+              } ) -> ()
+          (* TODO also skip variable declarations if they do not have
+                  any assignments *)
+          | (loc, _) -> Flow_js.add_warning cx [mk_reason "" loc, msg]
```

Comment:
The case for variable declarations is slightly trickier, because a single `VariableDeclaration` statement can have multiple actual declarations. You want to figure out if there is a `VariableDeclaration` where any of the `declarations` have a non-None `init` member. You could use a pattern guard for this.